### PR TITLE
[EraVM][Disassembler] Implement basic support for instruction decoding

### PR DESCRIPTION
**Pull request by atrosinenko**
Implement support for decoding simple instructions - those only using register and immediate operands.

Support for the instructions with in_any and out_any operands requires custom handling of stack operands and will be added via subsequent commits.

### CODE REVIEW

File: llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp [hunk 1]
```diff
@@ -39,8 +39,55 @@ class EraVMDisassembler : public MCDisassembler {
...
 };
+
+Register decodeGR256(const MCDisassembler *D, unsigned EncodedReg) {
+  const auto *MCRI = D->getContext().getRegisterInfo();
+  const auto &RC = MCRI->getRegClass(EraVM::GR256RegClassID);
+  assert(RC.getNumRegs() == 16 && "Unexpected change to register class");
+  return RC.getRegister(EncodedReg);
+}
+
+int64_t decodeCC(unsigned EncodedCC) {
```
Comment (atrosinenko) [inline llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:50]: Will update this to leverage #502.

Comment (atrosinenko) [reply llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:50]: Updated via 0916a7bc56339f35e37ed6a76a68d03ae0f6361c

**Comment by asl** [comment]
@atrosinenko Please rebase

File: llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp [hunk 2]
```diff
@@ -39,8 +39,53 @@ class EraVMDisassembler : public MCDisassembler {
...
+  const unsigned EncodingForIfGTOrLT = 7;
```
Comment (akiramenai) [inline llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:51]: Don't we have it defined somewhere? Also if we plan to support `GTOrLT` it makes sense to add `TODO: <issue number>`.

**Review by akiramenai — APPROVED** [review]
LGTM

**Comment by akiramenai** [comment]
We don't use `gsl` so I'm going to disable array to pointer decay checker.

File: llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp (hunk 2 above)
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:51]: At now, I assume `IfGTOrLT` to be deprecated, please correct me if I'm wrong. Related: #440.

Comment (akiramenai) [reply llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:51]: We probably didn't pay enough attention to this particular issue as it was discussed in context of other abbreviations.
I didn't hear about plans to remove it nor I can imagine a reason for doing so: we'd use 7 out of 8 possible predicates without it.
I would name it `gl`. :)
@sayon @hedgar2017 could you confirm?

Comment (hedgar2017) [reply llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:51]: Does is stand for `greater than or less than`? Why not name it `not equal` (`ne`)? Sounds more conventional.

Comment (akiramenai) [reply llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:51]: @hedgar2017 that why I proposed to rename flags to `x`, `y`, `z` to remove these associations. `gt`, `lt`, `eq` is a lie. It's true for `sub` instruction only. Instructions set some flags on overflow, on division by zero, etc. So `ne` means that `eq` flag is not set for this particular instruction; `gl` means that `gt` or `lt` is set regardless of whether `eq` is set (and if you ever see it in asm generated by our backend, gl with that)

Comment (akiramenai) [reply llvm/lib/Target/EraVM/Disassembler/EraVMDisassembler.cpp:51]: Keeping this conversation unresolved until we decide on #440.

**Pull request merged**